src/main/java/org/apache/commons/lang3/time/DateUtils.java

```diff
@@ -208,601 +217,759 @@ public static boolean isSameInstant(Date date1, Date date2) {
+    }
+
+    /**
+     * <p>Gets a date ceiling, leaving the field specified as the most
+     * significant field.</p>
+     *
+     * <p>For example, if you had the date-time of 28 Mar 2002
+     * 13:45:01.231, if you passed with HOUR, it would return 28 Mar
+     * 2002 14:00:00.000.  If this was passed with MONTH, it would
+     * return 1 Apr 2002 0:00:00.000.</p>
+     *
+     * @param date  the date to work with, not null
+     * @param field  the field from {@code Calendar} or {@code SEMI_MONTH}
+     * @return the different ceil date, not null
+     * @throws NullPointerException if the date is {@code null}
+     * @throws ArithmeticException if the year is over 280 million
+     * @since 2.5
+     */
+    public static Calendar ceiling(final Calendar date, final int field) {
+        if (date == null) {
+            throw nullDateIllegalArgumentException();
+        }
+        final Calendar ceiled = (Calendar) date.clone();
+        modify(ceiled, field, ModifyType.CEILING);
+        return ceiled;
+    }
+
+    /**
+     * <p>Gets a date ceiling, leaving the field specified as the most
+     * significant field.</p>
+     *
+     * <p>For example, if you had the date-time of 28 Mar 2002
+     * 13:45:01.231, if you passed with HOUR, it would return 28 Mar
+     * 2002 14:00:00.000.  If this was passed with MONTH, it would
+     * return 1 Apr 2002 0:00:00.000.</p>
+     *
+     * @param date  the date to work with, either {@code Date} or {@code Calendar}, not null
+     * @param field  the field from {@code Calendar} or {@code SEMI_MONTH}
+     * @return the different ceil date, not null
+     * @throws NullPointerException if the date is {@code null}
+     * @throws ClassCastException if the object type is not a {@code Date} or {@code Calendar}
+     * @throws ArithmeticException if the year is over 280 million
+     * @since 2.5
+     */
+    public static Date ceiling(final Object date, final int field) {
+        if (date == null) {
+            throw nullDateIllegalArgumentException();
+        }
+        if (date instanceof Date) {
+            return ceiling((Date) date, field);
+        }
+        if (date instanceof Calendar) {
+            return ceiling((Calendar) date, field).getTime();
+        }
+        throw new ClassCastException("Could not find ceiling of for type: " + date.getClass());
     }
 
-    //-----------------------------------------------------------------------
     /**
      * <p>Internal calculation method.</p>
-     * 
-     * @param val  the calendar
+     *
+     * @param val  the calendar, not null
      * @param field  the field constant
-     * @param round  true to round, false to truncate
+     * @param modType  type to truncate, round or ceiling
      * @throws ArithmeticException if the year is over 280 million
      */
-    private static void modify(Calendar val, int field, boolean round) {
+    private static void modify(final Calendar val, final int field, final ModifyType modType) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 31. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 948-980 into a separate method.